Tasks/Common/azure-arm-rest/azure-arm-common.ts

```diff
@@ -11,19 +12,24 @@ export class ApplicationTokenCredentials {
     public authorityUrl: string;
     public activeDirectoryResourceId: string;
     public isAzureStackEnvironment: boolean;
+    public scheme: number;
+    public msiClientId: string;
     private token_deferred: Q.Promise<string>;
 
-    constructor(clientId: string, domain: string, secret: string, baseUrl: string, authorityUrl: string, activeDirectoryResourceId: string, isAzureStackEnvironment: boolean) {
-        if (!Boolean(clientId) || typeof clientId.valueOf() !== 'string') {
-            throw new Error(tl.loc("ClientIdCannotBeEmpty"));
-        }
+    constructor(clientId: string, domain: string, secret: string, baseUrl: string, authorityUrl: string, activeDirectoryResourceId: string, isAzureStackEnvironment: boolean, scheme?: string, msiClientId?: string) {
 
         if (!Boolean(domain) || typeof domain.valueOf() !== 'string') {
             throw new Error(tl.loc("DomainCannotBeEmpty"));
         }
 
-        if (!Boolean(secret) || typeof secret.valueOf() !== 'string') {
-            throw new Error(tl.loc("SecretCannotBeEmpty"));
+        if((!scheme ||scheme ==='ServicePrincipal')){
```

Comment:
scheme != 'MSI' ?

Reply:
It looks more promising to check scheme === serviceprincipal and take action instead of scheme != MSI. As in future we need not change this block in case if we add new scheme.